swarm/runtime/statsdb/ingestion.py

```
@@ -552,25 +552,73 @@ def ingest_events(self, events: List[Dict[str, Any]], run_id: str) -> int:
 
-        for event in events:
-            # Ensure run_id is set on the event for raw storage
-            event_with_run = {**event, "run_id": run_id}
+        # Check which events already exist (in chunks to avoid large IN clauses)
+        existing_ids = set()
+        chunk_size = 1000
+        for i in range(0, len(event_ids), chunk_size):
+            chunk = event_ids[i : i + chunk_size]
+            placeholders = ",".join(["?"] * len(chunk))
+            rows = self.connection.execute(
+                f"SELECT event_id FROM events WHERE event_id IN ({placeholders})",
```

Ruff (0.15.0)

[error] Possible SQL injection vector through string-based query construction

(S608)